File: app/src/main/java/com/breadwallet/tools/util/TokenUtil.java
/**
 * BreadWallet
 * <p/>
 * Created by Jade Byfield <jade@breadwallet.com> on 9/13/2018.
 * Copyright (c) 2018 breadwallet LLC
 * <p/>
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * <p/>
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * <p/>
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.breadwallet.tools.util;

import android.annotation.SuppressLint;
import android.content.Context;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.core.util.Preconditions;

import com.breadwallet.BuildConfig;
import com.breadwallet.R;
import com.breadwallet.model.TokenItem;
import com.platform.APIClient;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import okhttp3.Request;

import static androidx.core.util.Preconditions.checkNotNull;

public final class TokenUtil {

    private static final String TAG = TokenUtil.class.getSimpleName();

    private static final String ENDPOINT_CURRENCIES = "/currencies";
    private static final String ENDPOINT_CURRENCIES_SALE_ADDRESS = "/currencies?saleAddress=";
    private static final String FIELD_CODE = "code";
    private static final String FIELD_NAME = "name";
    private static final String FIELD_SCALE = "scale";
    private static final String FIELD_CONTRACT_ADDRESS = "contract_address";
    private static final String FIELD_IS_SUPPORTED = "is_supported";
    private static final String FIELD_SALE_ADDRESS = "sale_address";
    private static final String FIELD_CONTRACT_INITIAL_VALUE = "contract_initial_value";
    private static final String FIELD_COLORS = "colors";
    private static final String FIELD_CURRENCY_ID = "currency_id";
    private static final String FIELD_TYPE = "type";
    private static final String FIELD_ALTERNATE_NAMES = "alternate_names";
    private static final String FIELD_CRYPTOCOMPARE = "cryptocompare";
    private static final String ICON_DIRECTORY_NAME_WHITE_NO_BACKGROUND = "white-no-bg";
    private static final String ICON_DIRECTORY_NAME_WHITE_SQUARE_BACKGROUND = "white-square-bg";
    private static final String ICON_FILE_NAME_FORMAT = "%s.png";
    private static final int START_COLOR_INDEX = 0;
    private static final int END_COLOR_INDEX = 1;
    private static final String TOKENS_FILENAME = "tokens.json";

    private static final String ETHEREUM = "ethereum";
    private static final String ETHEREUM_TESTNET = "ropsten";
    private static final String TESTNET = "testnet";
    private static final String MAINNET = "mainnet";

    // TODO: In DROID-878 fix this so we don't have to store this mTokenItems... (Should be stored in appropriate wallet.)
    private static List<TokenItem> mTokenItems = new ArrayList<>();
    private static Map<String, TokenItem> mTokenMap = new HashMap<>();

    private TokenUtil() {
    }

    /**
     * When the app first starts, fetch our local copy of tokens.json from the resource folder
     *
     * @param context The Context of the caller
     */
    public static void initialize(Context context, Boolean forceLoad) {
        String filePath = context.getFilesDir().getAbsolutePath() + File.separator + TOKENS_FILENAME;
        File tokensFile = new File(filePath);

        if (!tokensFile.exists() || forceLoad) {
            InputStream tokensInputStream = context.getResources().openRawResource(R.raw.tokens);
            BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(tokensInputStream));
            StringBuilder stringBuilder = new StringBuilder();
            String line;

            try {
                while ((line = bufferedReader.readLine()) != null) {
                    stringBuilder.append(line).append('\n');
                }

                // Copy the APK tokens.json to a file on internal storage
                saveTokenListToFile(context, stringBuilder.toString());
                loadTokens(parseJsonToTokenList(context, stringBuilder.toString()));
            } catch (IOException e) {
                Log.e(TAG, "Could not read from resource file at res/raw/tokens.json ", e);
            }
        }
    }

    /**
     * This method can either fetch the list of supported tokens, or fetch a specific token by saleAddress
     * Request the list of tokens we support from the /currencies endpoint
     *
     * @param context  The Context of the caller
     * @param tokenUrl The URL of the endpoint to get the token metadata from.
     */
    private static APIClient.BRResponse fetchTokensFromServer(Context context, String tokenUrl) {
        Request request = new Request.Builder()
                .url(tokenUrl)
                .header(BRConstants.HEADER_CONTENT_TYPE, BRConstants.CONTENT_TYPE_JSON_CHARSET_UTF8)
                .header(BRConstants.HEADER_ACCEPT, BRConstants.CONTENT_TYPE_JSON).get().build();

        return APIClient.getInstance(context).sendRequest(request, true);
    }

    /**
     * This method fetches a specific token by saleAddress
     *
     * @param context     The Context of the caller
     * @param saleAddress Optional sale address value if we are looking for a specific token response.
     */
    public static TokenItem getTokenItem(Context context, String saleAddress) {
        APIClient.BRResponse response = fetchTokensFromServer(context,
                APIClient.getBaseURL() + ENDPOINT_CURRENCIES_SALE_ADDRESS + saleAddress);
        if (response != null && !response.getBodyText().isEmpty()) {
            ArrayList<TokenItem> tokenItems = parseJsonToTokenList(context, response.getBodyText());

            // The response in this case should contain exactly 1 token item.
            return tokenItems == null || tokenItems.size() != 1 ? null : tokenItems.get(0);
        }

        return null;
    }

    public static synchronized List<TokenItem> getTokenItems(Context context) {
        if (mTokenItems == null || mTokenItems.isEmpty()) {
            loadTokens(getTokensFromFile(context));
        }
        return mTokenItems;
    }

    /**
     * Return a TokenItem with the given currency code or null if non TokenItem has the currency code.
     *
     * @param currencyCode The currency code of the token we are looking.
     * @return The TokenItem with the given currency code or null.
     */
    public static TokenItem getTokenItemByCurrencyCode(String currencyCode) {
        return mTokenMap.get(currencyCode.toLowerCase());
    }

    public static void fetchTokensFromServer(Context context) {
        APIClient.BRResponse response = fetchTokensFromServer(context,
                APIClient.getBaseURL() + ENDPOINT_CURRENCIES);

        if (response != null && !response.getBodyText().isEmpty()) {
            // Synchronize on the class object since getTokenItems is static and also synchronizes on the class object
            // rather than on an instance of the class.
            synchronized (TokenItem.class) {
                String responseBody = response.getBodyText();

                // Check if the response from the server is valid JSON before trying to save & parse.
                if (Utils.isValidJSON(responseBody)) {
                    saveTokenListToFile(context, responseBody);
                    loadTokens(parseJsonToTokenList(context, responseBody));
                }
            }
        } else {
            Log.e(TAG, "failed to fetch tokens: " + response.getCode());
        }
    }

    // TODO refactor to avoid passing in context.
    private static ArrayList<TokenItem> parseJsonToTokenList(Context context, String jsonString) {
        ArrayList<TokenItem> tokenItems = new ArrayList<>();
        // Iterate over the token list and announce each token to Core.
        try {
            JSONArray tokenListArray = new JSONArray(jsonString);

            for (int i = 0; i < tokenListArray.length(); i++) {
                JSONObject tokenObject = tokenListArray.getJSONObject(i);
                String address = "";
                String name = "";
                String symbol = "";
                String type = "";
                boolean isSupported = true;
                String cryptocompareAlias = null;

                if (tokenObject.has(FIELD_CONTRACT_ADDRESS)) {
                    address = tokenObject.getString(FIELD_CONTRACT_ADDRESS);
                }

                if (tokenObject.has(FIELD_NAME)) {
                    name = tokenObject.getString(FIELD_NAME);
                }

                if (tokenObject.has(FIELD_CODE)) {
                    symbol = tokenObject.getString(FIELD_CODE);
                }

                if (tokenObject.has(FIELD_IS_SUPPORTED)) {
                    isSupported = tokenObject.getBoolean(FIELD_IS_SUPPORTED);
                }

                if (tokenObject.has(FIELD_ALTERNATE_NAMES)) {
                    JSONObject altNameObject = tokenObject.getJSONObject(FIELD_ALTERNATE_NAMES);
                    if (altNameObject.has(FIELD_CRYPTOCOMPARE)) {
                        cryptocompareAlias = altNameObject.getString(FIELD_CRYPTOCOMPARE);
                    }
                }

                if (tokenObject.has(FIELD_TYPE)) {
                    type = tokenObject.getString(FIELD_TYPE);
                }

                if (!Utils.isNullOrEmpty(name) && !Utils.isNullOrEmpty(symbol)) {
                    String startColor = null;
                    String endColor = null;
                    String currencyId = "";
                    if (tokenObject.has(FIELD_COLORS)) {
                        JSONArray colorsArray = tokenObject.getJSONArray(FIELD_COLORS);
                        startColor = (String) colorsArray.get(START_COLOR_INDEX);
                        endColor = (String) colorsArray.get(END_COLOR_INDEX);
                    }

                    if (tokenObject.has(FIELD_CURRENCY_ID)) {
                        currencyId = tokenObject.getString(FIELD_CURRENCY_ID);

                        if (BuildConfig.BITCOIN_TESTNET) {
                            currencyId = currencyId.replace(MAINNET, currencyId.contains(ETHEREUM) ? ETHEREUM_TESTNET : TESTNET);
                        }
                    }

                    TokenItem item = new TokenItem(address, symbol, name, null, isSupported, currencyId, type, startColor, endColor, cryptocompareAlias);

                    tokenItems.add(item);
                }
            }
        } catch (JSONException e) {
            Log.e(TAG, "Error parsing token list response from server:", e);
        }
        return tokenItems;
    }

    private static void saveTokenListToFile(Context context, String jsonResponse) {
        String filePath = context.getFilesDir().getAbsolutePath() + File.separator + TOKENS_FILENAME;
        try {
            FileWriter fileWriter = new FileWriter(filePath);
            fileWriter.write(jsonResponse);
            fileWriter.flush();
            fileWriter.close();
        } catch (IOException e) {
            Log.e(TAG, "Error writing tokens JSON response to tokens.json:", e);
        }
    }

    private static List<TokenItem> getTokensFromFile(Context context) {
        try {
            File tokensFile = new File(context.getFilesDir().getPath() + File.separator + TOKENS_FILENAME);
            FileInputStream fileInputStream = new FileInputStream(tokensFile);
            int size = fileInputStream.available();
            byte[] fileBytes = new byte[size];
            fileInputStream.read(fileBytes);
            fileInputStream.close();

            return parseJsonToTokenList(context, new String(fileBytes));

        } catch (IOException e) {
            Log.e(TAG, "Error reading tokens.json file: ", e);
            return mTokenItems;
        }
    }

    public static String getTokenIconPath(Context context, String currencyCode, boolean withBackground) {
        String bundleResource = ServerBundlesHelper
                .getExtractedPath(context, ServerBundlesHelper.getBundle(ServerBundlesHelper.Type.TOKEN), null);
        String iconDirectoryName = withBackground
                ? ICON_DIRECTORY_NAME_WHITE_SQUARE_BACKGROUND : ICON_DIRECTORY_NAME_WHITE_NO_BACKGROUND;
        String iconFileName = String.format(ICON_FILE_NAME_FORMAT, currencyCode);
        File bundleDirectory = new File(bundleResource);
        if (bundleDirectory.exists() && bundleDirectory.isDirectory()) {
            for (File iconDirectory : bundleDirectory.listFiles()) {
                if (iconDirectory.getName().equalsIgnoreCase(iconDirectoryName) && iconDirectory.isDirectory()) {
                    for (File iconFile : iconDirectory.listFiles()) {
                        if (iconFile.getName().equalsIgnoreCase(iconFileName)) {
                            return iconFile.getAbsolutePath();
                        }
                    }
                }
            }
        }

        return "";
    }

    @SuppressLint("ResourceType")
    public static String getTokenStartColor(String currencyCode, Context context) {
        TokenItem tokenItem = mTokenMap.get(currencyCode.toLowerCase());
        if (tokenItem != null && !tokenItem.getStartColor().isEmpty()) {
            return tokenItem.getStartColor();
        } else {
            return context.getString(R.color.wallet_delisted_token_background);
        }
    }

    @SuppressLint("ResourceType")
    public static String getTokenEndColor(String currencyCode, Context context) {
        TokenItem tokenItem = mTokenMap.get(currencyCode.toLowerCase());
        if (tokenItem != null && !tokenItem.getEndColor().isEmpty()) {
            return tokenItem.getEndColor();
        } else {
            return context.getString(R.color.wallet_delisted_token_background);
        }
    }

    public static boolean isTokenSupported(String symbol) {
        if (mTokenMap.containsKey(symbol.toLowerCase())) {
            return mTokenMap.get(symbol.toLowerCase()).isSupported();
        }
        return true;
    }

    /**
     * Returns the currency code to be used when fetching the exchange rate for the token with the
     * given currency code.
     *
     * @param currencyCode the currency code
     * @return
     */
    public static String getExchangeRateCode(String currencyCode) {
        return mTokenMap.containsKey(currencyCode.toLowerCase())
                ? mTokenMap.get(currencyCode.toLowerCase()).getExchangeRateCurrencyCode() : currencyCode;
    }

    private static void loadTokens(List<TokenItem> tokenItems) {
        mTokenItems = tokenItems;
        mTokenMap.clear();
        for (TokenItem tokenItem : tokenItems) {
            mTokenMap.put(tokenItem.getSymbol().toLowerCase(), tokenItem);
        }
    }

    @Nullable
    public static TokenItem getTokenItemForCurrencyId(String currencyId) {
        checkNotNull(currencyId);
        for (TokenItem item : mTokenItems) {
            if (item.getCurrencyId().equalsIgnoreCase(currencyId)) {
                return item;
            }
        }
        return null;
    }
}

File: app-core/src/main/java/com/breadwallet/tools/util/TokenUtil.kt
/**
 * BreadWallet
 *
 * Created by Jade Byfield <jade@breadwallet.com> on 9/13/2018.
 * Copyright (c) 2018 breadwallet LLC
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package com.breadwallet.tools.util

import android.content.Context
import com.breadwallet.appcore.BuildConfig
import com.breadwallet.logger.logError
import com.breadwallet.model.TokenItem
import com.breadwallet.theme.R
import com.breadwallet.tools.manager.BRReportsManager
import com.platform.APIClient.BRResponse
import com.platform.APIClient.Companion.getBaseURL
import com.platform.APIClient.Companion.getInstance
import com.platform.util.getBooleanOrDefault
import com.platform.util.getJSONArrayOrNull
import com.platform.util.getJSONObjectOrNull
import com.platform.util.getStringOrNull
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import okhttp3.Request
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.File
import java.io.IOException
import java.util.ArrayList
import java.util.HashMap
import java.util.Locale
import kotlin.properties.Delegates

object TokenUtil {
    private const val ENDPOINT_CURRENCIES = "/currencies"
    private const val ENDPOINT_CURRENCIES_SALE_ADDRESS = "/currencies?saleAddress="
    private const val FIELD_CODE = "code"
    private const val FIELD_NAME = "name"
    private const val FIELD_SCALE = "scale"
    private const val FIELD_CONTRACT_ADDRESS = "contract_address"
    private const val FIELD_IS_SUPPORTED = "is_supported"
    private const val FIELD_SALE_ADDRESS = "sale_address"
    private const val FIELD_CONTRACT_INITIAL_VALUE = "contract_initial_value"
    private const val FIELD_COLORS = "colors"
    private const val FIELD_CURRENCY_ID = "currency_id"
    private const val FIELD_TYPE = "type"
    private const val FIELD_ALTERNATE_NAMES = "alternate_names"
    private const val FIELD_COINGECK_ID = "coingecko"
    private const val ICON_DIRECTORY_NAME_WHITE_NO_BACKGROUND = "white-no-bg"
    private const val ICON_DIRECTORY_NAME_WHITE_SQUARE_BACKGROUND = "white-square-bg"
    private const val ICON_FILE_NAME_FORMAT = "%s.png"
    private const val START_COLOR_INDEX = 0
    private const val END_COLOR_INDEX = 1
    private const val TOKENS_FILENAME = "tokens.json"
    private const val TOKENS_FILENAME_TESTNET = "tokens-testnet.json"
    private const val ETHEREUM = "ethereum"
    private const val ETHEREUM_TESTNET = "ropsten"
    private const val TESTNET = "testnet"
    private const val MAINNET = "mainnet"

    private var isMainnet: Boolean by Delegates.notNull()
    private lateinit var context: Context
    private var tokenItems: List<TokenItem> = ArrayList()
    private var tokenMap: Map<String, TokenItem> = HashMap()
    private val initLock = Mutex(locked = true)

    suspend fun waitUntilInitialized() = initLock.withLock { Unit }

    /**
     * When the app first starts, fetch our local copy of tokens.json from the resource folder
     *
     * @param context The Context of the caller
     */
    fun initialize(context: Context, forceLoad: Boolean, isMainnet: Boolean) {
        TokenUtil.isMainnet = isMainnet
        TokenUtil.context = context
        val fileName = if (isMainnet) TOKENS_FILENAME else TOKENS_FILENAME_TESTNET
        val tokensFile = File(context.filesDir, fileName)
        if (!tokensFile.exists() || forceLoad) {
            try {
                initLock.tryLock()
                val tokens = context.resources
                    .openRawResource(
                        if (isMainnet) {
                            com.breadwallet.appcore.R.raw.tokens
                        } else {
                            com.breadwallet.appcore.R.raw.tokens_testnet
                        }
                    ).reader().use { it.readText() }

                // Copy the APK tokens.json to a file on internal storage
                saveDataToFile(context, tokens, fileName)
                loadTokens(parseJsonToTokenList(tokens))
                initLock.unlock()
            } catch (e: IOException) {
                BRReportsManager.error("Failed to read res/raw/tokens.json", e)
            }
        } else {
            initLock.tryLock()
            fetchTokensFromServer()
            initLock.unlock()
        }
    }

    /**
     * This method can either fetch the list of supported tokens, or fetch a specific token by saleAddress
     * Request the list of tokens we support from the /currencies endpoint
     *
     * @param tokenUrl The URL of the endpoint to get the token metadata from.
     */
    private fun fetchTokensFromServer(tokenUrl: String): BRResponse {
        val request = Request.Builder()
            .get()
            .url(tokenUrl)
            .header(BRConstants.HEADER_CONTENT_TYPE, BRConstants.CONTENT_TYPE_JSON_CHARSET_UTF8)
            .header(BRConstants.HEADER_ACCEPT, BRConstants.CONTENT_TYPE_JSON)
            .build()
        return getInstance(context).sendRequest(request, true)
    }

    @Synchronized
    fun getTokenItems(): List<TokenItem> {
        if (tokenItems.isEmpty()) {
            loadTokens(getTokensFromFile())
        }
        return tokenItems
    }

    /**
     * Return a TokenItem with the given currency code or null if non TokenItem has the currency code.
     *
     * @param currencyCode The currency code of the token we are looking.
     * @return The TokenItem with the given currency code or null.
     */
    fun tokenForCode(currencyCode: String): TokenItem? {
        return tokenMap[currencyCode.toLowerCase(Locale.ROOT)]
    }

    fun tokenForCoingeckoId(coingeckoId: String): TokenItem? {
        val matches = coingeckoIdMap.filterValues { coingeckoId.equals(it, true) }
        return matches.keys.firstOrNull()?.run(TokenUtil::tokenForCode)
            ?: tokenItems.firstOrNull {
                coingeckoId.equals(it.coingeckoId, true)
            }
    }

    fun tokenForCurrencyId(currencyId: String): TokenItem? {
        return tokenItems.find { it.currencyId.equals(currencyId, true) }
    }

    fun coingeckoIdForCode(code: String): String? {
        return coingeckoIdMap[code.toUpperCase(Locale.ROOT)]
            ?: tokenForCode(code)?.coingeckoId
    }

    private fun fetchTokensFromServer() {
        if (!isMainnet) return
        val response = fetchTokensFromServer(getBaseURL() + ENDPOINT_CURRENCIES)
        if (response.isSuccessful && response.bodyText.isNotEmpty()) {
            // Synchronize on the class object since getTokenItems is static and also synchronizes
            // on the class object rather than on an instance of the class.
            synchronized(TokenItem::class.java) {
                val responseBody = response.bodyText

                // Check if the response from the server is valid JSON before trying to save & parse.
                if (Utils.isValidJSON(responseBody)) {
                    saveDataToFile(context, responseBody, TOKENS_FILENAME)
                    loadTokens(parseJsonToTokenList(responseBody))
                }
            }
        } else {
            logError("failed to fetch tokens: ${response.code}")
        }
    }

    private fun parseJsonToTokenList(jsonString: String): ArrayList<TokenItem> {
        val tokenJsonArray = try {
            JSONArray(jsonString)
        } catch (e: JSONException) {
            BRReportsManager.error("Failed to parse Token list JSON.", e)
            JSONArray()
        }
        return List(tokenJsonArray.length()) { i ->
            try {
                tokenJsonArray.getJSONObject(i).asTokenItem()
            } catch (e: JSONException) {
                BRReportsManager.error("Failed to parse Token JSON.", e)
                null
            }
        }.filterNotNull().run(::ArrayList)
    }

    private fun saveDataToFile(context: Context, jsonResponse: String, fileName: String) {
        try {
            File(context.filesDir.absolutePath, fileName).writeText(jsonResponse)
        } catch (e: IOException) {
            BRReportsManager.error("Failed to write tokens.json file", e)
        }
    }

    private fun getTokensFromFile(): List<TokenItem> = try {
        val fileName = if (isMainnet) TOKENS_FILENAME else TOKENS_FILENAME_TESTNET
        val file = File(context.filesDir.path, fileName)
        parseJsonToTokenList(file.readText())
    } catch (e: IOException) {
        BRReportsManager.error("Failed to read tokens.json file", e)
        tokenItems
    }

    fun getTokenIconPath(currencyCode: String, withBackground: Boolean): String? {
        val bundleResource = ServerBundlesHelper
            .getExtractedPath(
                context,
                ServerBundlesHelper.getBundle(ServerBundlesHelper.Type.TOKEN),
                null
            )
        val iconFileName = ICON_FILE_NAME_FORMAT.format(currencyCode.toLowerCase(Locale.ROOT))
        val iconDirectoryName = if (withBackground) {
            ICON_DIRECTORY_NAME_WHITE_SQUARE_BACKGROUND
        } else {
            ICON_DIRECTORY_NAME_WHITE_NO_BACKGROUND
        }
        val iconDir = File(bundleResource, iconDirectoryName)
        val iconFile = File(iconDir, iconFileName)
        return if (iconFile.exists()) iconFile.absolutePath else null
    }

    fun getTokenStartColor(currencyCode: String): String? {
        val tokenItem = tokenMap[currencyCode.toLowerCase(Locale.ROOT)]
        return if (tokenItem != null && !tokenItem.startColor.isNullOrBlank()) {
            tokenItem.startColor
        } else {
            context.getString(R.color.wallet_delisted_token_background)
        }
    }

    fun getTokenEndColor(currencyCode: String): String? {
        val tokenItem = tokenMap[currencyCode.toLowerCase(Locale.ROOT)]
        return if (tokenItem != null && !tokenItem.endColor.isNullOrBlank()) {
            tokenItem.endColor
        } else {
            context.getString(R.color.wallet_delisted_token_background)
        }
    }

    fun isTokenSupported(symbol: String): Boolean {
        return tokenMap[symbol.toLowerCase(Locale.ROOT)]?.isSupported ?: true
    }

    private fun loadTokens(tokenItems: List<TokenItem>) {
        val native = tokenItems.filter(TokenItem::isNative).sortedBy { it.name }
        val tokens = tokenItems.filterNot(TokenItem::isNative).sortedBy { it.symbol }
        TokenUtil.tokenItems = native + tokens
        tokenMap = TokenUtil.tokenItems.associateBy { item ->
            item.symbol.toLowerCase(Locale.ROOT)
        }
    }

    private fun JSONObject.asTokenItem(): TokenItem? = try {
        val (startColor, endColor) = getJSONArrayOrNull(FIELD_COLORS)?.run {
            getStringOrNull(START_COLOR_INDEX) to getStringOrNull(END_COLOR_INDEX)
        } ?: null to null

        val name = getString(FIELD_NAME)
        TokenItem(
            address = getStringOrNull(FIELD_CONTRACT_ADDRESS),
            symbol = getString(FIELD_CODE),
            name = name,
            image = null,
            isSupported = getBooleanOrDefault(FIELD_IS_SUPPORTED, true),
            currencyId = getString(FIELD_CURRENCY_ID),
            type = getString(FIELD_TYPE),
            startColor = startColor,
            endColor = endColor,
            coingeckoId = getJSONObjectOrNull(FIELD_ALTERNATE_NAMES)
                ?.getStringOrNull(FIELD_COINGECK_ID)
        )
    } catch (e: JSONException) {
        BRReportsManager.error("Token JSON: $this")
        BRReportsManager.error("Failed to create TokenItem from JSON.", e)
        null
    }
}
